### PR TITLE
Verdaccio script and CI build w/ verdaccio packages

### PR DESCRIPTION
**Pull request by LFDanLu**
Closes https://github.com/adobe/react-spectrum/issues/2389

## ✅ Pull Request Checklist:

- [ ] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [ ] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [ ] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [ ] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## 🧢 Your Project:

RSP

### CODE REVIEW

**Comment by LFDanLu** [comment]
Test app now has a bunch more components (but not all, I'll add more in the future if we are cool with the setup)

**Comment by LFDanLu** [comment]
@ktabors 
> Not sure if this running on the master branch in CI is the best. By that point it is too late because the docs change was deployed.

Actually it would be ok since the docs changes that get deployed are built off the currently published packages so the verdaccio built docs would show what the docs WILL look like upon the next release. This is because the currently published packages are behind what is on main since main hasn't been released yet

> It wasn't obvious from this PR how to run it locally, but that ended up being easy, ./scripts/verdacio.sh in one command line tab as running process (localhost:4000) and make website-production in a second command line tab once verdacio is up. You'll need to run a python (python -m SimpleHTTPServer 8080) or node http server (http-server) from the docs directory (dist/production/docs).

Haha, good point and thanks for testing it with me on your local machine. If we are ok with this script I'll go ahead and add some documentation somewhere with the same steps you listed above

**Comment by LFDanLu** [comment]
Looks like something broke when attempting to build the docs w/ verdaccio since the last release, will need to dig

File: scripts/verdaccio.sh
Comment (LFDanLu) [inline scripts/verdaccio.sh:91]: For now I'm just storing the size of the test build into a text file that then is uploaded to our azure storage. I've taken a look at things like https://packtracker.io/ for a place to host this data instead but I'm unsure if this integration is something I need to check w/ security first (org perms required?).

Alternatively, I could try comparing the app size with the app size in the last publish commit, need to poke around for the feasibility of this.

File: .circleci/comment.js
Comment (devongovett) [inline .circleci/comment.js:45]: Is this why I've been getting lots of random notifications on commits? Is this going to be annoying?

I think it might not only happen on main also, because you can commit to branches that aren't in PRs.

File: scripts/compareSize.js
Comment (devongovett) [inline scripts/compareSize.js:33]: How are you planning on keeping this updated?

File: scripts/verdaccio.sh
Comment (devongovett) [inline scripts/verdaccio.sh:38]: Maybe use `make build` here so we avoid duplication?

Also `@adobe/react-spectrum` shouldn't be built by Parcel.

File: scripts/verdaccio.sh
Comment (devongovett) [inline scripts/verdaccio.sh:2]: Not sure, but you might want `set -e` here? That way if any command fails, we don't continue with the following ones? Or does the `trap` below handle that?

File: .circleci/comment.js
Comment (devongovett) [inline .circleci/comment.js:55]: Also I see these weren't added to the below PR commenting logic. So does this mean the comments only go on commits and not on PRs?

File: .circleci/comment.js
Comment (LFDanLu) [reply .circleci/comment.js:45]: Ah, lemme go ahead and add a more strict check for main. I'm surprised you've been getting notifications on your commits, shouldn't this change be isolated to this branch?

The ideal scenario is that this only runs on commits in main, open to suggestions on the desired cadence. I think it would be nice if it ran only on the commit before publish/release but wasn't able to think up of a way to do so

File: .circleci/comment.js
Comment (LFDanLu) [reply .circleci/comment.js:55]: That is correct, I figured we'd only really want to do the verdaccio builds on commits to main since it is quite time consuming and I'd rather not hold up PR merges

File: scripts/compareSize.js
Comment (LFDanLu) [reply scripts/compareSize.js:33]: I'll be deleting this fallback after we get this merged and a release is done. That way the script can always look up the commit before the last release and use that as the "lastApp/lastPublish" data point. The hardcoded fallback is only here so we can have an initial data point of comparison

File: scripts/verdaccio.sh
Comment (LFDanLu) [reply scripts/verdaccio.sh:2]: Good tip, I'll set that just in case the errexit behavior bails out if a command fails (it seems to have already been behaving like this in my experience but looking it up seems to indicate that it is false by default)
The `trap` below handles running the cleanup function upon receiving the `ERR` or `EXIT` signal.

File: scripts/verdaccio.sh
Comment (LFDanLu) [reply scripts/verdaccio.sh:38]: Hrm, if we don't build `@adobe/react-spectrum` how do I get it to publish to verdaccio for consumption in the test app?

File: .circleci/comment.js
Comment (devongovett) [reply .circleci/comment.js:45]: No, not on my commits. I think maybe they were the commits on this PR. But they link to a commit rather than a PR.

File: scripts/verdaccio.sh
Comment (devongovett) [reply scripts/verdaccio.sh:38]: The mono packages have a `prepublishOnly` script that runs as part of publishing.

File: .circleci/comment.js
Comment (devongovett) [reply .circleci/comment.js:55]: Ah I see. Maybe we can chat about this with the team. I wonder if we might want to know before merge whether we broke the prod docs build or something. But we can see how it goes.

File: .circleci/comment.js
Comment (LFDanLu) [reply .circleci/comment.js:45]: Ah right, it was spammy because I was doing some frequent commits to handle issues between ci vs local differences, apologies about that. When we merge and have it run on commits to main only it shouldn't be as spammy (at least no less spammy than a standard PR with its comment per commit push)

File: scripts/verdaccio.sh
Comment (LFDanLu) [reply scripts/verdaccio.sh:38]: Hm, interesting, those prepublish scripts should be running via this script as well, lemme try it again

Comment (devongovett) [reply scripts/verdaccio.sh:38]: It shouldn't be any different than our usual nightly or release publishing steps I think

Comment (LFDanLu) [reply scripts/verdaccio.sh]: yep, confirmed that `make build` works locally, not sure why I saw otherwise when helping with the test app PR...

**Review by ktabors — COMMENTED** [review]
Is the only stat we are looking at size(example project size and RSP size)?
Is it only building on main right now? I can’t see a verdacio built docs or storybook in the last build comment of the PR? I thought you could get it to build on this branch.
Why did you have to switch node versions from 14 to 16?

**Comment by LFDanLu** [comment]
@ktabors 
- yep for now we look at the build project size and the size of each deployed package. Open to other stats/metrics but those should suffice for now IMO
- I changed it to build on main only in preparation for an eventual merge to main. Take a look at https://github.com/adobe/react-spectrum/commit/379c5bf2c95125623510953aa834fca0a449b7ef#comments for the build comments. For reference, you can click the speech button icon on the commits: 
image
- I had to bump the node versions for installing/building example project app if I remember correctly

**Review by ktabors — DISMISSED** [review]

File: scripts/buildWebsite.js
Comment (devongovett) [inline scripts/buildWebsite.js:62]: Why did we add `--no-scope-hoist` here?

Comment (LFDanLu) [reply scripts/buildWebsite.js:62]: Ah, at the time `--no-scope-hoist` was already a part of this cmd line call from main. I see it isn't there anymore on main so I'll remove as well

**Review by devongovett — APPROVED** [review]

**Review by reidbarber — APPROVED** [review]
This is awesome!

File: .circleci/config.yml
Comment (snowystinger) [inline .circleci/config.yml:451]: I think it's problematic for this to have the same key as the one above it? why don't they conflict?

File: examples/rsp-cra-18/package.json
Comment (snowystinger) [inline examples/rsp-cra-18/package.json:8]: latest for workflow as well?

File: .circleci/config.yml
Comment (LFDanLu) [reply .circleci/config.yml:451]: The `name` provided to the job handles distinguishing the two jobs even though they use the same overall `comment` job defined above https://github.com/adobe/react-spectrum/blob/f8b270c1da9f19c8d19583fefcd8f44671226506/.circleci/config.yml#L345-L357

Results in a flow like so: https://app.circleci.com/pipelines/github/adobe/react-spectrum/9885/workflows/9ca28065-70c7-4494-9f9b-639838077f23

File: examples/rsp-cra-18/package.json
Comment (LFDanLu) [reply examples/rsp-cra-18/package.json:8]: ah yep good idea, that I can bump in the next iteration

**Pull request merged**
